extensions/amp-analytics/0.1/amp-analytics-3p-message-queue.js

```diff
@@ -0,0 +1,168 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {dev} from '../../../src/log';
+import {AMP_ANALYTICS_3P_MESSAGE_TYPE} from '../../../src/3p-analytics-common';
+import {SubscriptionApi} from '../../../src/iframe-helper';
+
+/** @private @const {string} */
+const TAG_ = 'amp-analytics.3pMessageQueue';
+
+/** @private @const {number} */
+const MAX_QUEUE_SIZE_ = 100;
+
+/**
+ * @visibleForTesting
+ */
+export class AmpAnalytics3pMessageQueue {
+  /**
+   * Constructor
+   * @param {!Window} win The window element
+   * @param {!HTMLIFrameElement} frame The cross-domain iframe to send
+   * messages to
+   */
+  constructor(win, frame) {
+    /** @private {!Window} */
+    this.win_ = win;
+
+    /** @private {!HTMLIFrameElement} */
+    this.frame_ = frame;
+
+    /** @private {boolean} */
+    this.isReady_ = false;
+
+    /** @private {!../../../src/3p-analytics-common.AmpAnalytics3pEvent} */
+    this.creativeToPendingMessages_ = {};
+
+    /** @private
+     *  {!../../../src/3p-analytics-common.AmpAnalytics3pNewCreative} */
+    this.creativeToExtraData_ = {};
```

Comment:
I removed the class that contained it, but after speaking with Hui, we do still need to send the extraData.